DEAD feat(nx-plugin): add NX Plugin with OpenAPI client generator and executor

seriouslag opened the pull request:
This NX plugin will generate an NX package from an OpenAPI spec file and run the openapi-ts on the spec.

This is useful for NX projects.
Also very useful to let AI agents know how to use the plugin to automate setting up client code to fit into an NX project.

TODO:
- [x] Add NX generator
- [x] Accept openapi spec input, dereference and bundle the spec file using functions from openapi-ts
- [x] Setup tests using vitest
- [x] Add an executor to update the API spec file and regenerate the client code
- [x] Get build working (facing issues with integrating into monorepo)
- [x] Make plugins configurable and add the ability to add additional openapi-ts plugins
- [x] Support additional client types
- [ ] List package in NX plugin repository

Stretch in this PR:
- [ ] Support different test frameworks
- [ ] Setup linting support
- [ ] Auto symlink the package if in workspace mode?
- [ ] Auto-run the sync executor after generating a new package?


Questions:
Q: How to handle when the OpenAPI spec file updates?
A: We have added an executor and a call to each generated project. Each project can run that executor to update the spec file and regenerate the client code.

Q: How to we list the plugin in the NX plugin repository
A: [https://nx.dev/extending-nx/recipes/publish-plugin#list-your-nx-plugin](https://nx.dev/extending-nx/recipes/publish-plugin#list-your-nx-plugin)



Related PRs:
- [Update Vitest](https://github.com/hey-api/openapi-ts/pull/1934)
- [Export common functions to dogfood capabilities](https://github.com/hey-api/openapi-ts/pull/1934)

seriouslag committed:
feat(nx-plugin): add OpenAPI client generator and related files

- Introduced a new Nx plugin for generating OpenAPI client libraries from OpenAPI spec files.
- Added configuration files, README, and TypeScript setup for the generator.
- Implemented functionality to bundle and dereference OpenAPI specs using Redocly CLI.
- Included tests for the generator functionality and project structure validation.
- Updated package dependencies and added necessary scripts for project management.

seriouslag committed:
fix(nx-plugin): temporarily disable build dependency in test configuration

- Commented out the "dependsOn" field in the test configuration to prevent test failures due to the current build issues.

seriouslag committed:
test(nx-plugin): update axios client type in OpenAPI client generator tests

seriouslag committed:
feat(nx-plugin): implement update-api executor for OpenAPI spec management

- Added a new executor to update the OpenAPI specification and regenerate the client code based on changes.

seriouslag committed:
refactor(nx-plugin): update executor and generator paths and descriptions

- Changed implementation paths for the update-api executor and openapi-client generator to include 'index.js'.

seriouslag committed:
feat(nx-plugin): use built in dependency management and update paths for windows

seriouslag committed:
feat(nx-plugin): enhance OpenAPI client generator and update executor

- Added support for plugins in the OpenAPI client generator and update-api executor.
- Updated paths and constants for better maintainability and cross-platform compatibility.
- Introduced new dependencies and updated package versions in the configuration.
- Improved test cases to accommodate the new plugins feature.

seriouslag committed:
feat(nx-plugin): add support for different clients and plugins

seriouslag committed:
fix(nx-plugin): fix swagger v2 support

seriouslag committed:
fix(nx-plugin): update package name format in OpenAPI client generator template

seriouslag committed:
fix(nx-plugin): update default client type in OpenAPI client generator README

seriouslag committed:
refactor(nx-plugin): change execSync of @hey-api cli to js library

- this change provides better typings

seriouslag committed:
feat(nx-plugin): enhance OpenAPI client generator with test support and improved options

seriouslag committed:
test(nx-plugin):update tests to prevent race conditions

seriouslag committed:
fix(nx-plugin): remove jest config from tsconfig exclusion and lint template

seriouslag committed:
fix(nx-plugin): remove prompt for test runner in OpenAPI client schema

seriouslag committed:
fix(nx-plugin): update test generator adding tsconfig includes

seriouslag committed:
fix(nx-plugin): improve error handling and logging in update-api executor

seriouslag committed:
test(nx-plugin): mock latest-version in OpenAPI client generator tests

seriouslag committed:
fix(nx-plugin): enhance error handling in updatePackageJson function for tsconfig updates

seriouslag commented on packages/nx-plugin/src/executors/update-api/index.ts:
```diff
@@ -0,0 +1,198 @@
+      } else {
+        logger.debug('No existing spec file found. Creating...');
+      }
+      await writeFile(existingSpecPath, newSpecString);
```
Any ideas on how to handle this?

mrlubos replied:
The link explains how to fix the issue, but I'm not sure I'd focus on it. My understanding is the described risk doesn't apply here and it's okay to ignore

seriouslag committed:
chore(.gitignore): add timestamp files for vitest and vite configurations

seriouslag committed:
feat(vitest): add workspace configuration for Vitest to support multiple config file types

seriouslag committed:
test(nx-plugin): correct OpenAPI spec file comparisons

- the incorrect files were sent to the comparison checks
- fixed tests to actually handle temp folders

seriouslag committed:
feat(nx-plugin): add force option to update executor and move compare spec

seriouslag committed:
chore(dependencies): lock find-my-way to version 9.2.0 because they introduced a breaking change

seriouslag committed:
chore(nx-plugin): remove private flag and add prepublish script to package.json

seriouslag committed:
feat(openapi-ts): export getSpec function

Exporting the getSpec function to reuse the functionality in other packages

seriouslag committed:
refactor(nx-plugin): remove need to external package to read spec files in favor of dog fooding functions from openapi-ts

seriouslag committed:
refactor(nx-plugin): removed redocly in favor of dogfooding openapi-ts

seriouslag committed:
refactor(nx-plugin): replace async file writes with synchronous counterparts

seriouslag committed:
refactor(nx-plugin): add nx caching and test tsconfig

seriouslag committed:
fix(nx-plugin): ensure unique tags in normalizeOptions function

seriouslag committed:
feat(nx-plugin): enhance logging

seriouslag committed:
feat(vitest): streamline workspace configuration

seriouslag committed:
refactor(openapi-client): remove hanging test check

seriouslag committed:
fix(nx-plugin): ensure unique temporary directories in tests using random UUIDs

seriouslag committed:
fix(nx-plugin): improve logging and ensure unique temporary directories in update-api executor

seriouslag committed:
test(nx-plugin): skip tests

seriouslag committed:
fix(plugin): rebase

seriouslag committed:
chore(deps): update dependencies and improve client initialization in examples
